lib/charms/smtp_integrator/v0/smtp.py

```diff
@@ -0,0 +1,286 @@
+    port: int = Field(None, ge=1, le=65536)
+    user: Optional[str]
+    password: Optional[str]
+    auth_type: AuthType
+    transport_security: TransportSecurity
+    domain: Optional[str]
+
+    def to_relation_data(self) -> dict[str, str]:
+        """Convert an instance of SmtpRelationData to the relation representation.
+
+        Returns:
+            Dict containing the representation.
+        """
+        result = {
+            "host": str(self.host),
+            "port": str(self.port),
+            "auth_type": self.auth_type.value,
+            "transport_security": self.transport_security.value,
+        }
+        if self.domain:
+            result["domain"] = self.domain
+        if self.user:
+            result["user"] = self.user
+        if self.password:
+            result["password"] = self.password
+        return result
+
+
+class SmtpDataAvailableEvent(ops.RelationEvent):
+    """Smtp event emitted when relation data has changed.
+
+    Attrs:
+        host: The hostname or IP address of the outgoing SMTP relay.
+        port: The port of the outgoing SMTP relay.
+        user: The SMTP AUTH user to use for the outgoing SMTP relay.
+        password: The SMTP AUTH password to use for the outgoing SMTP relay.
+        auth_type: The type used to authenticate with the SMTP relay.
+        transport_security: The security protocol to use for the outgoing SMTP relay.
+        domain: The domain used by the sent emails from SMTP relay.
+    """
+
+    @property
+    def host(self) -> str:
+        """Fetch the SMTP host from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("host")
+
+    @property
+    def port(self) -> int:
+        """Fetch the SMTP port from the relation."""
+        assert self.relation.app
+        return int(self.relation.data[self.relation.app].get("port"))
+
+    @property
+    def user(self) -> str:
+        """Fetch the SMTP user from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("user")
+
+    @property
+    def password(self) -> str:
+        """Fetch the SMTP password from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("password")
+
+    @property
+    def auth_type(self) -> str:
+        """Fetch the SMTP auth type from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("auth_type")
+
+    @property
+    def transport_security(self) -> str:
+        """Fetch the SMTP transport security protocol from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("transport_security")
+
+    @property
+    def domain(self) -> str:
+        """Fetch the SMTP domain from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("domain")
```

Comment:
@gregory-schiano and @benhoyt as well for thoughts, these properties have the possibility of raising `AssertionError` (not sure under what conditions `self.relation.app` could be non-truthy). Additionally, the `.get` functions could return `None` which the type hints doesn't seem to indicate. Any thoughts on that?

Reply:
Perhaps the easiest way to solve this would be to make these functions and raise some sort of custom error if `self.relation.app` is falsy or when the `.get` returns `None`

Reply:
It looks like `Relation.app` should "basically never" be None. I know that's not very convincing :-) ... but the code tries very hard to set that. I've opened an [issue](https://github.com/canonical/operator/issues/1040) to investigate further and see if we can change the type annotation for `Relation.app` to be straight `Application` (that is, non-optional).

As the type annotation stands, however, I think the `assert` is exactly right: we're telling the type checker (and Python) we are sure this shouldn't be None.

A helper function `relation_app_data` that does this sounds reasonable though.

Reply:
If it is just aimed at type hints then [`typing.cast`](https://docs.python.org/3/library/typing.html#typing.cast) might be better because it can't raise an assertion error

It would be good if the type could be changed to be non optional if it can basically never be `None` 🙂

Reply:
> It would be good if the type could be changed to be non optional if it can basically never be None 🙂

Agreed. We'll aim to do that.

Reply:
This assertion doesn't bother me too much, raising an assertion error is not ideal but allows us to catch cases that shouldn't happen which is the goal of asserts in such kind of code.
`typing.cast` is also a valid approach